### PR TITLE
 binance: parseOrder: can't guess cost of filled order if no filled price

### PR DESCRIPTION
**Pull request by mkutny**

### CODE REVIEW

**Comment by kroitor** [comment]
I was thinking that setting `cost = filled * price` there unconditionally would be better for partial fills. Both `filled` and `price` are always defined. Otherwise we won't get any cost at all up until it is fully closed... what would you say?

**Comment by mkutny** [comment]
It's great that you mentioned it!

Actually, I've found the `cost` ambiguous and searched for it the Manual first.

It's not there so I was about to initiate discussion and submit a PR with Manual update first but was in hurry, get confused, and submitted this one a bit out of order. So let it hang on until we decide.

Ok, `cost` now.

As I see there are two perspectives on it:

- When an order is open I'd like to know how 'heavy' it is in terms of `price` * `remaining`, e.g. how much will it cost me to fulfill it. Similarly to what think before order creation: 'how much will it cost?', 'do I have funds available?', 'how big it is in comparison with other open orders?`. When an order is open I think of the future and in terms of `remaining`.

- When an order is closed I'd like to know how much it actually cost and what was the average price. With this regard, I think of the past and in terms of `filled`.

So, likewise, we have `filled` and `remaining` for the `amount`, I feel like we also need to have realized and remaining parts for `cost`. One `cost` won't please everybody.


> This, in fact, maybe a useless `cost`, because the `price` is not the actual filling price, we should get that actual price from trades anyway.

What you say actually makes a lot of sense. How about the following:
- remaining cost we take from open orders (`price` * `remaining`)
- realized/spent cost we take from the trades
- if order is closed/canceled the `cost` is `undefined`

**Comment by kroitor** [comment]
> As I see there are two perspectives on it

I would add a third one, the reverse of the first:
- When an order is still open, but has been partially filled (already accumulated some cost) – we would like to know how much exactly was reported from the exchange, to make sure we are properly aligned with their/our current fee schedule without having to actually fetch the trades.

In general, I agree with all of what you said, we might need several values for costs. Like initial: { cost, amount, price }, filled: { cost, amount, price }, remaining: { cost, amount, price }... Apart from the costs we would like to have multiple timestamps, namely, the creation time and the last change time / closing time.

**Comment by mkutny** [comment]
> we would like to know how much exactly was reported from the exchange, to make sure we are properly aligned with their/our current fee schedule

Just out of curiosity (I'm honestly curious):
- when might this be needed? I saw you were concerned with fees before and I'm just curious what type of trading would require knowing fees in the realtime. Let's say I distribute a huge inventory and the order takes hours to fulfill. The only case when actual fees are different from expected is when I aimed to place a maker order but it hit another order by chance and it partly become a taker. Ok, I know that I missed. But from now on I can do nothing about it, those extra fees have gone, right? What am I missing?
- fees are *a part* of the cost, aren't they? how can I looking at the cost tell the fees?


>  initial: { cost, amount, price }, filled: { cost, amount, price }, remaining: { cost, amount, price }

Ok, to summarise:
- initial cost = amount * price = remaining * price
- remaining cost = remaining * price (the same as above)
- filled cost = filled * average price (if available) or undefined

So basically two costs are enough and we can proceed without breaking changes by adding just one property for filled cost and one for the average price. How would you feel about that?

**Comment by kroitor** [comment]
> when might this be needed? I saw you were concerned with fees before and I'm just curious what type of trading would require knowing fees in the realtime. Let's say I distribute a huge inventory and the order takes hours to fulfill. The only case when actual fees are different from expected is when I aimed to place a maker order but it hit another order by chance and it partly become a taker. Ok, I know that I missed. But from now on I can do nothing about it, those extra fees have gone, right? What am I missing?

In general, there are two cases:
1. We don't live in an ideal world. But we still want some $) So, we should expect (be prepared for) all sorts of inconsistencies from exchanges themselves. Check this out: https://github.com/ccxt/ccxt/issues/1084. I bet the entire OK family has the same engine, and you were recently active on okcoin, so you might be interested.
2. Whenever we encounter some unexpected order cost or fee cost, we are triggering or scheduling a corresponding reaction. I can't really tell you all of it, but if you think of it at scale, you'll be interested in ways of handling miniscule losses in batches.

> So basically two costs are enough and we can proceed without breaking changes by adding just one property for filled cost and one for the average price. How would you feel about that?

I would say, the real meaning behind the `initial` cost is `expected` cost. It's there just for convenience (to save one line in userland). We can drop it, if you think it isn' necessary, it is always equal to price * amount. We keep it because some exchanges like HitBTC won't return full order info, when you fetch those orders and parse them later.

**Comment by mkutny** [comment]
Ok, agree to drop `initial` cost.
`remaining` is actually the same.

There is uncertainty with `filled` cost because it might or might not equal to `average price` * `amount` and is a tool for the discrepancy detection.

So the `cost` means the `filled cost`. As a result we need to set it to `undefined` for all other cases, otherwise, users won't be able to tell them apart, right?

**Comment by kroitor** [comment]
> Ok, agree to drop initial cost.

I was talking on the opposite – keep the initial, as long as it does not break anything, for convenience and, also, because of the following reason:

> remaining is actually the same.

Nope, it is not the same. I mean, remaining says how much is left to buy, so this value changes. But the initial cost stays the same. Let's call it "total cost as expected on our side". I would keep both values (with proper renaming). 

> There is uncertainty with filled cost because it might or might not equal to average price * amount and is a tool for the discrepancy detection.

I believe, the filled cost should not be an average. It should be an exact value from the exchange (where known). The average cost does not make too much sense for a practical purpose, I think we always want the exact actual filled cost, yet, I might be missing something here.

> So the cost means the filled cost. As a result we need to set it to undefined for all other cases, otherwise, users won't be able to tell them apart, right?

It depends on the point of view. There are two "views":
1. Current state of the library
2. The future reworked state of the library (not existing yet)

If we are talking of 1, currently there is just the cost, which takes best available value, an exact one where available, or an average one, or the precalculated one. So, yes, there is an inconsistency here that we have to deal with, it is, of course, subject to change.

From the perspective of 2, we should, I think, devise some spec on it that will account for the above. Namely, we should have:
- total expected cost of the entire order
- exact cost of currently partially-filled volume
- expected cost of the rest of the volume to be filled

And the same properties should probably exist for amounts, prices and fees. It will most likely require a redesign of some parts of order management logic. The first thing we would like to do, is to build a robust order cache from trade fills log (it's a work in progress, more about it here: https://github.com/ccxt/ccxt/issues/569).

**Comment by Rutinguer** [comment]
Im using this for resolve MIN NOTIONAL problem
